dynawo/sources/Solvers/Common/DYNSolverImpl.cpp

```diff
@@ -245,73 +245,54 @@ Solver::Impl::solve(double tAim, double &tNxt) {
 bool
 Solver::Impl::evalZMode(vector<state_g> &G0, vector<state_g> &G1, const double & time) {
   Timer timer("SolverIMPL::evalZMode");
-  bool zChange = false;
-  bool modeChange = false;
-  bool stableRoot = true;
   bool change = false;
 
-  for (int i = 0; i < maxNumberUnstableRoots; ++i) {
-    // evalZ
+  // evalZ part
+  bool nonSilentZChange;
+  int i = 0;
+  do {
+    nonSilentZChange = false;
     model_->evalZ(time);
-    zChange = model_->zChange();
     ++stats_.nze_;
 
-    // evaluate G and compare with previous values
-    stableRoot = detectUnstableRoot(G0, G1, time);
-
-    if (model_->getSilentZChange())
+    if (model_->getSilentZChange()) {
       state_.setFlags(SilentZChange);
-    if (zChange) {
-      change = true;
+    } else if (model_->zChange()) {
```

Comment:
I would remove this else if as it relies on a assumption of the behavior of the model (silentZ flag is not set whenever zChange is set) that might become false in the future

Reply:
If I understand it correctly, you suggest to have:

if (model_->getSilentZ()) {
blabla
} else {
bloublou
}. 

The problem for me, at least in theory, is that there could be a g change leading to the call to evalZMode that won't have any effect on z but only on the modes. In this case, if there is no silentZChange neither zChange, I will be stuck in the loop and even if I write another exit condition, I would have one non necessary call to evalG.

One example should be an equation like this if Q < QMax then Q else QMax for example.

Reply:
Let me try to be clearer: I was proposing to remove the "else":
 if (model_->getSilentZChange()) {
      state_.setFlags(SilentZChange);
if (model_->zChange()) {
...
}

So that if model_->zChange() is true we systematically call evalG independently from the silentZ flag

Reply:
Ok, and not removing the "if": one chance out of two.
Fine then!

Reply:
I think I finally get it! I haven't in mind and didn't understand that zChange and silentZChange were complementary (one model can't have both).

Reply:
Solved